### ISSUE TITLE
Enhancement: [naming-convention] Accept Promise type

### ISSUE THREAD

**Issue by Smtih** [typescript-eslint/typescript-eslint#5107]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md

### Description

I've seen that it is not possible to allow any non-primitive type name, however given we have rules like `no-floating-promises` it seems possible to detect if something is at the very least Promise-like.

I would love to enforce some naming conventions on promises as often it's not obvious that an async function may not have been executed yet (especially when returning the promise at the end of the function which means it's not detected by `no-floating-promises`)

```json
"@typescript-eslint/naming-convention": [
    "error",
    {
      "selector": "variable",
      "types": ["Promise"],
      "format": ["snakeCase"],
      "suffix": ["Promise"]
    }
  ]
```

### Fail

```typescript
const person = getPersonAsync();
```

### Pass

```typescript
const personPromise = getPersonAsync();
```

**Comment by bradzacher** [comment]
Unfortunately there is no way to tell if something is a promise.

Instead rules like `no-floating-promise` do [duck-typing](https://en.m.wikipedia.org/wiki/Duck_typing) checks on types to guess ths they are promises (eg "does it have a `.then` and a `.catch`? Probably a promise").
So at best we can answer "is the thing roughly, maybe, potentially assignable to the `PromiseLike` interface?".

This works "well enough" for those rules, but is known to cause issues when working with things that use similar APIs.

---

The reason the rule supports the things that it does is because they are all concrete, tangible things in the type system that we can unequivocally look for due to them being primitives.
Because we cannot have the same certainty about Promises, I don't think it's a good thing to add.

Additionally I don't think this request has broad applicability - the rule has been released for a while now and this is the first time this feature had been asked for.

Finally I have never seen someone use a special naming convention for promise-like variables.

For the above reasons, I am going to reject this proposal. Thanks!
